legacy/dockerregistry/inventory.go

```diff
@@ -1152,39 +1165,40 @@ func (sc *SyncContext) IgnoreFromPromotion(regName RegistryName) {
 // could be removed in a future promoter manifest version which could force the
 // user to provide these delineations for us.
 //
-// nolint[gocyclo]
-func ParseContainerParts(s string) (string, string, error) {
+// TODO: Can we simplify this to not use switch/case/goto?
+func ParseContainerParts(s string) (
+	registry string,
+	repo string,
+	parseErr error,
+) {
 	parts := strings.Split(s, "/")
 	if len(parts) <= 1 {
-		goto error
+		goto InvalidString
 	}
+
 	// String may not have a double slash, or a trailing slash (which would
 	// result in an empty substring).
 	for _, part := range parts {
-		if len(part) == 0 {
-			goto error
+		if part == "" {
+			goto InvalidString
 		}
 	}
+
 	switch parts[0] {
-	case "gcr.io":
-		fallthrough
-	case "asia.gcr.io":
-		fallthrough
-	case "eu.gcr.io":
-		fallthrough
-	case "us.gcr.io":
+	case "gcr.io", "asia.gcr.io", "eu.gcr.io", "us.gcr.io":
 		if len(parts) == 2 {
-			goto error
+			goto InvalidString
 		}
 		return strings.Join(parts[0:2], "/"), strings.Join(parts[2:], "/"), nil
-	case "k8s.gcr.io":
-		fallthrough
-	case "staging-k8s.gcr.io":
-		fallthrough
 	default:
+		if parts[0] != "k8s.gcr.io" && parts[0] != "staging-k8s.gcr.io" {
```

Comment:
Sorry I wasn't able to review this code.

I'd rather have the branchless case statements with fallthrough than have the 2 `!=` conditionals with `&&`. Can we avoid this and still call InvalidString if it's some unexpected string?

Reply:
@listx -- A lot of this is minimum viable lint fixes, as the amount of cleanup required was more than I was expecting.

I've left a `TODO` atop the function that someone can follow-up on:

```go
// TODO: Can we simplify this to not use switch/case/goto?
```

That said, branchless case statements were the reason for the lint warnings, so I don't think we should move backwards:

```console
legacy/dockerregistry/inventory.go:1191:3: emptyFallthrough: remove empty case containing only fallthrough to default case (gocritic)
		fallthrough
		^
legacy/dockerregistry/inventory.go:1189:3: emptyFallthrough: remove empty case containing only fallthrough to default case (gocritic)
		fallthrough
		^
legacy/dockerregistry/inventory.go:1182:3: emptyFallthrough: replace empty case containing only fallthrough with expression list (gocritic)
		fallthrough
		^
legacy/dockerregistry/inventory.go:1180:3: emptyFallthrough: replace empty case containing only fallthrough with expression list (gocritic)
		fallthrough
		^
legacy/dockerregistry/inventory.go:1178:3: emptyFallthrough: replace empty case containing only fallthrough with expression list (gocritic)
		fallthrough
		^
```

ref: https://github.com/kubernetes-sigs/k8s-container-image-promoter/pull/394#issuecomment-894833433

Reply:
If we're going to make the linter happy only to create a TODO comment, in the future I would advise to just turn off the linter locally at that location (and leave a TODO about why we're suppressing the linter). In this case the linter made our code harder to read :(

Reply:
The goal of the linting was not to create extraneous TODOs.

> just turn off the linter locally at that location (and leave a TODO about why we're suppressing the linter)

There are several places across this codebase where I do exactly this, but in this case the TODO was left to later assess the entirety of the function (which is why it was placed atop and not next to one of the `case`s).